File: server/Services/VaultsService.cs






namespace keeper.Services;
public class VaultsService
{
  public VaultsService(VaultsRepository vaultsRepository, ProfilesService profilesService)
  {
    _vaultsRepository = vaultsRepository;
    _profilesService = profilesService;
  }
  private readonly VaultsRepository _vaultsRepository;
  private readonly ProfilesService _profilesService;


  internal Vault CreateVault(Vault vaultData)
  {
    Vault vault = _vaultsRepository.CreateVault(vaultData);
    return vault;
  }

  private Vault GetVaultById(int vaultId)
  {
    Vault vault = _vaultsRepository.GetVaultById(vaultId);

    if (vault == null) throw new Exception($"Invalid vault id: {vaultId}");

    return vault;
  }

  internal Vault GetVaultById(int vaultId, string userId)
  {
    Vault vault = GetVaultById(vaultId);

    if (vault.CreatorId != userId && vault.IsPrivate == true) throw new Exception($"Invalid vault id: {vaultId} 😜");

    return vault;
  }

  internal Vault UpdateVault(int vaultId, string userId, Vault vaultData)
  {
    Vault vault = GetVaultById(vaultId);

    if (vault.CreatorId != userId) throw new Exception("YOU CANNOT ALTER ANOTHER USERS VAULT, MY LOVE");

    vault.Name = vaultData.Name ?? vault.Name;
    vault.Description = vaultData.Description ?? vault.Description;
    vault.Img = vaultData.Img ?? vault.Img;
    vault.IsPrivate = vaultData.IsPrivate ?? vault.IsPrivate;

    _vaultsRepository.UpdateVault(vault);
    return vault;
  }

  internal string DeleteVault(int vaultId, string userId)
  {
    Vault vault = GetVaultById(vaultId);

    if (vault.CreatorId != userId) throw new Exception("YOU CANNOT DELETE ANOTHER USERS VAULT, DUDE");

    _vaultsRepository.DeleteVault(vaultId);
    return $"Deleted {vault.Name}";
  }

  internal List<Vault> GetVaultsByProfileId(string profileId)
  {
    _profilesService.GetProfileById(profileId);
    List<Vault> vaults = GetMyVaults(profileId);

    return vaults.FindAll(vault => vault.CreatorId == profileId && vault.IsPrivate == false);
  }

  internal List<Vault> GetMyVaults(string userId)
  {
    List<Vault> vaults = _vaultsRepository.GetMyVaults(userId);
    return vaults;
  }


}

File: client/src/services/VaultsService.js
import { api } from "./AxiosService.js"
import { Vault } from "@/models/Vault.js"
import { AppState } from "@/AppState.js"

class VaultsService {
  async getVaultById(vaultId) {
    AppState.activeVault = null
    const response = await api.get(`api/vaults/${vaultId}`)
    const vault = new Vault(response.data)
    AppState.activeVault = vault
  }
  async deleteVault(vaultId) {
    const response = await api.delete(`api/vaults/${vaultId}`)
    const vaultIndex = AppState.vaults.findIndex(vault => vault.id == vaultId)
    AppState.vaults.splice(vaultIndex, 1)
  }
  async createVault(vaultData) {
    const response = await api.post('api/vaults', vaultData)
    const createdVault = new Vault(response.data)
    // if (vaultData.creatorId != AppState.activeProfile.id) return
    AppState.vaults.unshift(createdVault)
  }
  async getMyVaults() {
    const response = await api.get('account/vaults')
    const vaults = response.data.map(vault => new Vault(vault))
    AppState.vaults = vaults

  }
  async GetVaultsByProfileId(profileId) {
    const response = await api.get(`api/profiles/${profileId}/vaults`)
    const vaults = response.data.map(vault => new Vault(vault))
    AppState.vaults = vaults
  }

}

export const vaultsService = new VaultsService()